src/HTTPure/MultiHeaders.purs

```diff
@@ -0,0 +1,107 @@
+-- | such as "Set-Cookie".
+newtype MultiHeaders = MultiHeaders (Map CaseInsensitiveString (NonEmptyArray String))
+
+derive instance newtypeMultiHeaders :: Newtype MultiHeaders _
+
+derive instance genericMultiHeaders :: Generic MultiHeaders _
+
+-- | Given a string, return a `Maybe` containing the values of the matching
+-- | header, if there is any.
+instance lookupMultiHeaders :: Lookup MultiHeaders String (NonEmptyArray String) where
+  lookup (MultiHeaders headersMap) key = headersMap !! key
+
+instance showMultiHeaders :: Show MultiHeaders where
+  show = genericShow
+
+-- | Compare two `MultiHeaders` objects by comparing the underlying `Objects`.
+derive newtype instance eqMultiHeaders :: Eq MultiHeaders
+
+-- | Allow one `MultiHeaders` objects to be appended to another.
+instance semigroupMultiHeaders :: Semigroup MultiHeaders where
+  append (MultiHeaders a) (MultiHeaders b) =
+    MultiHeaders $ Data.Map.unionWith append a b
+
+-- | Return a `MultiHeaders` containing nothing.
+empty :: MultiHeaders
+empty = MultiHeaders Data.Map.empty
+
+-- | Create a singleton header from a key-value pair.
+header :: String -> String -> MultiHeaders
+header key = header' key <<< Data.Array.NonEmpty.singleton
+
+-- | Create a singleton header from a key-values pair.
+header' :: String -> NonEmptyArray String -> MultiHeaders
+header' key = MultiHeaders <<< Data.Map.singleton (CaseInsensitiveString key)
+
+-- | Convert an `Array` of `Tuples` of 2 `Strings` to a `MultiHeaders` object.
+headers :: Array (Tuple String String) -> MultiHeaders
+headers = headers' <<< map (map pure)
+
+-- | Convert an `Array` of `Tuples` of 2 `Strings` to a `MultiHeaders` object.
+headers' :: Array (Tuple String (NonEmptyArray String)) -> MultiHeaders
+headers' = foldl insertField Data.Map.empty >>> MultiHeaders
+  where
+  insertField x (Tuple key values) = Data.Map.insertWith append (CaseInsensitiveString key) values x
+
+-- | Read the headers out of a HTTP `Request` object and parse duplicated
+-- | headers as a list (instead of comma-separated values, as with
+-- | `HTTPure.Headers.read`).
+read :: Request -> MultiHeaders
+read = requestHeadersDistinct >>> fold insertField Data.Map.empty >>> MultiHeaders
+  where
+  insertField ::
+    forall a.
+    Map CaseInsensitiveString (NonEmptyArray a) ->
+    String ->
+    Array a ->
+    Map CaseInsensitiveString (NonEmptyArray a)
+  insertField headersMap key array = case Data.Array.NonEmpty.fromArray array of
+    Nothing -> headersMap
+    Just nonEmptyArray -> Data.Map.insert (CaseInsensitiveString key) nonEmptyArray headersMap
+
+  -- | Similar to `Node.HTTP.requestHeaders`, but there is no join logic and the
+  -- | values are always arrays of strings, even for headers received just once.
+  -- | See https://nodejs.org/api/http.html#messageheadersdistinct.
+  requestHeadersDistinct :: Request -> Object (Array String)
+  requestHeadersDistinct = _.headersDistinct <<< unsafeCoerce
```

Comment:
Ah, this is convenient. I was imagining we'd use FFI to only return the headers with an array value (leaving out those with a string value), but considering `message.headersDistinct` already exists, this seems like the way to go.